cpp/src/arrow/filesystem/s3fs_benchmark.cc

```diff
@@ -0,0 +1,344 @@
+    Aws::S3::Model::PutObjectRequest req;
+    req.SetBucket(ToAwsString(bucket_));
+    req.SetKey(ToAwsString(name));
+    req.SetBody(std::make_shared<std::stringstream>(std::string(size, 'a')));
+    return OutcomeToStatus(client_->PutObject(req));
+  }
+
+  /// Make an object with Parquet data.
+  Status MakeParquetObject(const std::string& path, int num_columns, int num_rows) {
+    std::vector<std::shared_ptr<ChunkedArray>> columns(num_columns);
+    std::vector<std::shared_ptr<Field>> fields(num_columns);
+    for (int i = 0; i < num_columns; ++i) {
+      arrow::random::RandomArrayGenerator generator(i);
+      std::shared_ptr<Array> values = generator.Float64(num_rows, -1.e10, 1e10, 0);
+      std::stringstream ss;
+      ss << "col" << i;
+      columns[i] = std::make_shared<ChunkedArray>(values);
+      fields[i] = ::arrow::field(ss.str(), values->type());
+    }
+    auto schema = std::make_shared<::arrow::Schema>(fields);
+
+    std::shared_ptr<Table> table = Table::Make(schema, columns);
+
+    std::shared_ptr<io::OutputStream> sink;
+    ARROW_ASSIGN_OR_RAISE(sink, fs_->OpenOutputStream(path));
+    auto writer_properties =
+        parquet::WriterProperties::Builder().write_batch_size(128)->build();
+    auto arrow_properties = parquet::default_arrow_writer_properties();
+    RETURN_NOT_OK(parquet::arrow::WriteTable(*table, arrow::default_memory_pool(), sink,
+                                             num_rows, writer_properties,
+                                             arrow_properties));
+
+    return Status::OK();
+  }
+
+  void TearDown(const ::benchmark::State& state) override { ASSERT_OK(minio_.Stop()); }
+
+ protected:
+  MinioTestServer minio_;
+  std::string region_;
+  std::string bucket_;
+  Aws::Client::ClientConfiguration client_config_;
+  Aws::Auth::AWSCredentials credentials_;
+  std::unique_ptr<Aws::S3::S3Client> client_;
+  S3Options options_;
+  std::shared_ptr<S3FileSystem> fs_;
+};
+
+/// Set up/tear down the AWS SDK globally.
+/// (GBenchmark doesn't run GTest environments.)
+class S3BenchmarkEnvironment {
+ public:
+  S3BenchmarkEnvironment() { s3_env->SetUp(); }
+  ~S3BenchmarkEnvironment() { s3_env->TearDown(); }
+};
+
+S3BenchmarkEnvironment env{};
+
+/// Read the entire file into memory in one go to measure bandwidth.
+static void NaiveRead(benchmark::State& st, S3FileSystem* fs, const std::string& path) {
+  int64_t total_bytes = 0;
+  int total_items = 0;
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    std::shared_ptr<Buffer> buf;
+    int64_t size;
+    ASSERT_OK_AND_ASSIGN(file, fs->OpenInputFile(path));
+    ASSERT_OK_AND_ASSIGN(size, file->GetSize());
+    ASSERT_OK_AND_ASSIGN(buf, file->ReadAt(0, size));
+    total_bytes += buf->size();
+    total_items += 1;
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
```

Comment:
`std::cerr` is used in other parts. Is there a reason for the discrepancy?

Reply:
An oversight on my part; I've made these all cerr.